torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1156,6 +1175,19 @@ struct to_ir {
         return RefinementSet({present}, {});
       }
     }
+    if (const auto& union_type = lhs_value->type()->cast<UnionType>()) {
+      std::vector<Refinement> present;
+      for (const auto& type : union_type->types()) {
```

Comment:
You're refining the name to each contained type that isn't None. What we want is to refine the type to its contained types minus NoneType.